able to ignore vendored files certain build tags

dqminh opened the pull request:
Some go packages includes files with certain build tags that is only
relevant for code generation and can usually be ignored when vendoring.
An example of such build tag is `+build ignore` as stated in go/build
doc:

```
To keep a file from being considered for the build:
// +build ignore
(any other unsatisfied word will work as well, but “ignore” is conventional.)
```

This patch add `-ignore-tags` flag which can be set as a list of tags to
be ignored when vendoring. This defaults to `[]string{"ignore"}`.

`IgnoreTags` will be processed before `UseAllFiles`, hence files
matching `IgnoreTags` will be matched even with `UseAllFiles` are set.
We also pruned such all IgnoredGoFiles as well.

LK4D4 commented:
@dqminh looks good, I have some insane logic in vndr itself to handle this, but your solution is way better.
Do you want to write a small test? I can try to do it if you have no time.
Thanks!

dqminh commented:
@LK4D4 updated with a test. Also the old code has some bugs regarding arbitrary tag names, i think i have fixed them. PTAL.

Running on some test repos like containerd and runs yields correct results i think:

For containerd:

```
        deleted:    vendor/github.com/Microsoft/hcsshim/mksyscall_windows.go
        deleted:    vendor/github.com/containerd/continuity/manifest_test_darwin.go
        deleted:    vendor/golang.org/x/sys/unix/mkpost.go
        deleted:    vendor/golang.org/x/sys/unix/types_darwin.go
        deleted:    vendor/golang.org/x/sys/unix/types_dragonfly.go
        deleted:    vendor/golang.org/x/sys/unix/types_freebsd.go
        deleted:    vendor/golang.org/x/sys/unix/types_linux.go
        deleted:    vendor/golang.org/x/sys/unix/types_netbsd.go
        deleted:    vendor/golang.org/x/sys/unix/types_openbsd.go
        deleted:    vendor/golang.org/x/sys/unix/types_solaris.go
```

For runc:

```
        deleted:    vendor/golang.org/x/sys/unix/mkpost.go
        deleted:    vendor/golang.org/x/sys/unix/types_darwin.go
        deleted:    vendor/golang.org/x/sys/unix/types_dragonfly.go
        deleted:    vendor/golang.org/x/sys/unix/types_freebsd.go
        deleted:    vendor/golang.org/x/sys/unix/types_netbsd.go
        deleted:    vendor/golang.org/x/sys/unix/types_openbsd.go
        deleted:    vendor/golang.org/x/sys/unix/types_solaris.go
```

LK4D4 commented:
@dqminh thanks! Everything looks good. Do you think we need this flag? Maybe just skip `ignore` by default for now and leave flag implementation + docs for an occasion when someone needs it.

dqminh commented:
@LK4D4 updated to remove the default value

LK4D4 commented:
@dqminh oh, I meant to remove flag at all. And always ignore `ignore` tag.

dqminh commented:
@LK4D4 removed the flag. PTAL.

LK4D4 commented:
@dqminh thanks!

thaJeztah commented:
LOL, I was wondering why CI failed on the `vendor` check. Found it :joy:

dqminh commented:
@thaJeztah ooops, sorry :man_facepalming:

On the other hand, it deletes 16k lines of code though :p

thaJeztah commented:
@dqminh yes! I love the improvement, we just forgot to "pin" `vndr` itself to a specific version, so I fixed that in https://github.com/docker/cli/pull/318, allowing us to have control over changes like this
